File: framework/src/play-ahc-ws/src/main/java/play/libs/ws/ahc/AhcWSComponents.java
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package play.libs.ws.ahc;

import play.Environment;
import play.api.libs.ws.ahc.AsyncHttpClientProvider;
import play.components.AkkaComponents;
import play.components.ConfigurationComponents;
import play.inject.ApplicationLifecycle;
import play.libs.ws.WSClient;
import play.shaded.ahc.org.asynchttpclient.AsyncHttpClient;

/**
 * AsyncHttpClient WS implementation components.
 *
 * <p>Usage:</p>
 *
 * <pre>
 * public class MyComponents extends BuiltInComponentsFromContext implements AhcWSComponents {
 *
 *   public MyComponents(ApplicationLoader.Context context) {
 *       super(context);
 *   }
 *
 *   // some service class that depends on WSClient
 *   public SomeService someService() {
 *       // wsClient is provided by AhcWSComponents
 *       return new SomeService(wsClient());
 *   }
 *
 *   // other methods
 * }
 * </pre>
 *
 * @see play.BuiltInComponents
 * @see WSClient
 */
public interface AhcWSComponents extends WSClientComponents, ConfigurationComponents, AkkaComponents {

    Environment environment();

    ApplicationLifecycle applicationLifecycle();

    default WSClient wsClient() {
        AsyncHttpClient asyncHttpClient = new AsyncHttpClientProvider(
            configuration(),
            environment().asScala(),
            applicationLifecycle().asScala()
        ).get();

        return new AhcWSClient(asyncHttpClient, materializer());
    }
}

File: transport/client/play-ahc-ws/src/main/scala/play/api/libs/ws/ahc/AhcWSComponents.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws.ahc

import scala.concurrent.ExecutionContext

import org.apache.pekko.stream.Materializer
import play.api._
import play.api.inject.ApplicationLifecycle
import play.api.libs.ws.StandaloneWSClient
import play.api.libs.ws.WSClient
import play.shaded.ahc.org.asynchttpclient.AsyncHttpClient

/**
 * AsyncHttpClient WS API implementation components.
 */
trait AhcWSComponents {
  def environment: Environment

  def configuration: Configuration

  def applicationLifecycle: ApplicationLifecycle

  def materializer: Materializer

  def executionContext: ExecutionContext

  lazy val wsClient: WSClient = {
    implicit val mat: Materializer = materializer
    new AhcWSClientProvider(standaloneWSClient.asInstanceOf[StandaloneAhcWSClient]).get
  }

  lazy val standaloneWSClient: StandaloneWSClient = {
    implicit val mat: Materializer = materializer
    new StandaloneAhcWSClient(asyncHttpClient)
  }

  lazy val asyncHttpClient: AsyncHttpClient = {
    implicit val ec: ExecutionContext = executionContext
    new AsyncHttpClientProvider(environment, configuration, applicationLifecycle).get
  }

}
